src/Nethermind/Nethermind.Synchronization/SnapSync/ProgressTracker.cs

```diff
@@ -471,5 +542,55 @@ public void Dispose()
                 range?.Dispose();
             }
         }
+
+        public interface IStorageRangeLock
+        {
+            IStorageRangeLock Increment(uint value = 1);
+            void Decrement(ValueHash256 key, bool removeFromOwner);
+            void ExecuteSafe(Action action);
+        }
+
+        public class StorageRangeLock(uint counter, ConcurrentDictionary<ValueHash256, IStorageRangeLock> owner)
+            : IStorageRangeLock
+        {
+            private readonly object _lock = new();
+
+            public IStorageRangeLock Increment(uint value = 1)
+            {
+                lock (_lock)
+                {
+                    Counter += value;
+                }
+                return this;
+            }
+
+            public void Decrement(ValueHash256 key, bool removeFromOwner)
+            {
+                lock (_lock)
+                {
+                    if (--Counter == 0 && removeFromOwner)
+                        owner.TryRemove(key, out _);
+                }
+            }
```

Comment:
Does counter need to be under same lock as safe execution?
If not better to do it on `Interlocked`